### DISCUSSION TITLE
Can you tell me what editing this assembly value did?

### DISCUSSION THREAD

**Discussion by SirEnder125** [RedShyGuy/Vapecord-ACNL-Plugin#41] [Q&A] [answered]
Hello. I used CTRPF search to find the address of gem apples in TKCD.
I then watched the value with GDB debugger, and then edited the  value.
Once I edited the value,  GDB told  me the  address of  the function  that
made that happen (I think). And it was `0x1e8ca0`.

I  used  CTRPF  hex  editor  to  edit  the  ARM  value  ("instruction"?)  at  this
address.  I don't know what `E1A00000`  means,  but  that's what I  changed
the  value to.  After changing  `0x1e8c98` through `0x1e8ca0` to  `E1A000000`,
it no longer changed my amount of gem apples when I bought something.
(Which is very useful, but I want to understand what I did.)

If you can tell me what I did exactly, that'd be great!

**Comment by RedShyGuy** [comment, marked as answer]
E1A00000 is the instruction `mov r0, r0` which moves the data from register 0 to register 0. Basically doing absolutely nothing. You could call it `nop` as that is the instruction name for something which actually does nothing (would be `E320F000` btw).

**Reply by SirEnder125** [reply]
Okay, thanks. One more thing that I don't understand, though.
Why do I have to edit all 3 values? (Those of the addresses
0x1e8c98 to 0x1e8ca0)

**Reply by RedShyGuy** [reply]
I don't know how the function looks like so I can't tell you why.

**Reply by SirEnder125** [reply]
Oh, of course. Sorry. Thanks for the help. :wink:

**Comment by SirEnder125** [comment]
Hello. I was wondering, if I were to find the address of a ASM line that edits a value,
could I make a hook on that line? And maybe print something every time that value
is edited? If so, how would you go about doing this?

**Reply by RedShyGuy** [reply]
I can send you an example later when I am home

**Reply by RedShyGuy** [reply]
So, there are multiple ways of approaching that. And it really depends where your hook is placed. 
How I like to do it the most is by doing this:
Inside your .s file:
```asm
.arm
.fpu vfp
.align(4)

.macro FUNCTION name
    .section .text.\name
    .global \name
    .type \name, %function
    .align 2
\name:
.endm

FUNCTION    TestFunc
    PUSH      {R0-R3, LR}
    MOV       R0, #0             @put the instruction here where you hooked at
    BL        __Notifier
    POP       {R0-R3, PC}
```
Inside your .cpp file:
```cpp
extern "C" void TestFunc(void);

extern "C" void __Notifier() { //you can put anything you want into here
    OSD::Notify("Test");
}

void SomeFunctionOfYours(void) {
       static Hook hook;
       hook.Initialize(AddressToHookAt, (u32)TestFunc);
       hook.SetFlags(USE_LR_TO_RETURN);
       hook.Enable();
}
```
This should work fine!

**Reply by SirEnder125** [reply]
Okay. Thanks a lot. Although, this looks very complicated. I will have to
learn a bit of arm ASM I guess. Lol.

**Reply by RedShyGuy** [reply]
It's very basic ASM, you don't have to learn much.

**Reply by SirEnder125** [reply]
I suppose not. I am just confused about all the backslashes and .arm, .align, .fpu stuff.
Also, one `.align` has parentheses and one does not. Also, what even is a `.align` for? 
That kind of stuff confuses me. Unless I can figure out what it means. I'll see if I can find out.

**Reply by RedShyGuy** [reply]
You don't need to understand that part, its just there so the compiler knows what that function below is and all

**Reply by SirEnder125** [reply]
Oh, yeah. Like in the code [here](https://github.com/rydoginator/ACNL-NTR-Cheats/blob/master/Sources/hooks.s).
Okay. I'll try to learn what the rest
does. Thanks!

**Reply by SirEnder125** [reply]
Oh, but what does this mean?
```ASM
@put the instruction here where you hooked at
```

**Reply by RedShyGuy** [reply]
When you hook at something there was an instruction originally, take that instruction and place it there

**Reply by SirEnder125** [reply]
Oh, okay. I'll try it out. Thanks a ton!

EDIT: Do you mean that I'd replace `MOV R0, #0` with the instruction of the address that I am hooking on?

**Reply by RedShyGuy** [reply]
Yes

**Reply by SirEnder125** [reply]
Okay. Thanks. :)

**Reply by SirEnder125** [reply]
I got it working! :)
I thought I had something wrong, but it turned out that I forgot to enable the cheat. xD

**Reply by RedShyGuy** [reply]
Awesome! 🙂

**Reply by SirEnder125** [reply]
It looks like something IS wrong. I meant for it to write a message when
the amount of money in my wallet changed, and it does, but it also does
it when I pull a weed... Any idea why this would happen? I put the hook
on `0x3036f8`.

**Reply by RedShyGuy** [reply]
Uuh thats weird. If I recall correctly that address belongs to the function which sets money. So it really shouldn't do anything with weed lol. But I can look at it tomorrow to make sure.

**Reply by SirEnder125** [reply]
Okay, thank you! It is indeed very weird.

**Reply by RedShyGuy** [reply]
Soo, I looked at it.
The function you placed your Hook at is the general Encryption function of New Leaf. Used for various things.
And I have no clue why but they encrypt the counter for how many weeds you plucked in total it seems lol. 
That's why it also notifies when you pick weed. 
To fix it just hook here for example: `0x2C02E0`

**Reply by SirEnder125** [reply]
Okay, thanks! I'll try hooking there. :D

**Reply by SirEnder125** [reply]
Works great! Thanks!!
